### PR TITLE
Fix connection

### PR DESCRIPTION
**Pull request by guy120494**
I have created a new connection (EnhancedConnection) that works with a special dataFetcher (PaginationDataFetcher). The connection calls the dataFetcher for the specific entities, asks for the cursor (so the implementer of PaginationDataFetcher is responsible for the cursors and hence knows when is the last/first page) and asks for hasPrev/Next page.

If you want to use the enhanced, you need to use 
`@GraphQLConnection(connection = EnhancedConnectionFetcher.class)`

(The default connection is still using SimpleList - for backward compatibility)

@tdraier - I will be happy if you could code review me

### CODE REVIEW

File: src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java
Comment (bbakerman) [inline src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java]: this seems a little weird in that you represent the "encoded cursors" back to the `paginationDataFetcher` that provided them.  Therefore to implement this one would need to "decode" the value it just encoded.  Not the end of the world but ...

Perhaps you need the `paginationDataFetcher` to pass back pagination info on the original fetch via a compound value intermediate value.

    class CompundData {
         List<T> data
         boolean hasNext
        boolean hasPrev
    }

File: src/main/java/graphql/annotations/connection/ConnectionFetcher.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/ConnectionFetcher.java]: I think it would be more clear that ConnectionFetcher<T> actually extends DataFetcher<Connection<T>> , so T should be the object type

File: src/main/java/graphql/annotations/connection/ListConnection.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/ListConnection.java]: ListConnection could be renamed ListConnectionFetcher and implement ConnectionFetcher : 
`public class ListConnectionFetcher<T> extends SimpleListConnection<T> implements ConnectionFetcher<T>`

File: src/main/java/graphql/annotations/connection/StreamConnection.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/StreamConnection.java]: StreamConnection could also be renamed StreamConnectionFetcher, and use generics : 
`public class StreamConnectionFetcher<T> implements DataFetcher<Connection<T>>, ConnectionFetcher<T>`

File: src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java
Comment (tdraier) [reply src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java]: I agree, in my proposal it would mean directly returning this compound object instead of the lists : 
`CompoundData <T> getAfterCursor(String, int)` , `CompoundData<T> getBeforeCursor(String, int)`, so the interface will only need 3 methods ( with getCursor(T) ), which would make it even more clear.

**Review by tdraier — CHANGES_REQUESTED** [review]
Hi, thanks for the proposal ! It's a great improvement, however I think we can simplify the way it's done. I don't like the fact that we need to use a custom DataFetcher to be able to use pagination. By defining a data fetcher we loose much simplicity of using graphql-java-annotations (need to move the logic in an additional class implementing standard graphql-java DataFetcher) - and there's also no easy/clear way to defined field argument (only by defining @GraphQLField on a method, which will never be called, not very nice). I think it would be great if we can achieve the same thing by using the provided DataFetcher (Method, or even Property/Field). What would you think of moving the PaginationDataFetcher to the field object itself ( so something like PaginatedList ), so doing something like : 

```
@GraphQLField
@GraphQLConnection
public PaginatedList<MyObject> getObjects(@GraphQLArgument param1, @GraphQLArgument param2) {
}
```

Or even : 

```
@GraphQLField
@GraphQLConnection
public PaginatedList<MyObject> objects;
```

This way, no custom DataFetcher or even ConnectionFetcher - the DispatchingConnectionFetcher could be used to detect that the return is a PaginatedList, and then dispatch to the EnhancedConnectionFetcher to create a correct Connection object (btw, it would even be better if the DispatchingConnectionFetcher could be configured so that new pairs of return types/connectionFetcher could be added). 

The code in `ConnectionDataFetcher` could be simplified again ( DispatchingConnectionFetcher will still be built with the "data" object - no specific case - only difference is that "data" is not a List or Stream of items, but a PaginatedList object ). The role of handling pagination arguments will be kept in the ConnectionFetcher (in your proposal it's the role of the custom data fetcher to handle the pagination arguments), which would be more consistent with the other ConnectionFetcher (ListConnection and StreamConnection). 

The PaginatedList interface may need additional method to support returning sublist based on cursor/length (`List<T> getAfterCursor(String, int)` , `List<T> getBeforeCursor(String, int)`. It will be the role of the PaginatedList implementor to not load all items at creation, but only fetch the sublist when requested.

Globally I think it would be more clear and easier to use .. to implement pagination on a connection, you just return a PaginatedList object instead of a List, and implement the pagination logic here. What do you think ?

File: src/main/java/graphql/annotations/connection/ConnectionFetcher.java
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/ConnectionFetcher.java]: I tried to do so, but since SimpleListConnection also extends `DataFetcher<Connection<T>>`, I got into some troubles. 

We could just get rid of SimpleListConnection and support only the EnhancedConnection

**Comment by guy120494** [comment]
What if we combine your suggestion with @bbakerman's?
The `PaginatedList` will have what Brad had in mind:
`List<T> data` `hasNextPage` `hasPreviousPage`. To get the cursors we can have a method `public Map<T,String> getCursors` (T is the entity's type)That way we don't need a custom dataFetcher, but the `PaginatedList` is simpler. That way the `ConnectionDataFetcher` knows how to build the connection properly (assuming that the datafetcher fetched the relevant data, using "first" "after" etc., so no need of specific methods).

The only problem I see is that SimpleListConnection will not work, but I'm more than fine to get rid of it, it will make things a lot easier

**Comment by tdraier** [comment]
Yes, it makes things even more simpler. Still I prefer `String getCursor(T)` over `Map<T,String> getCursors` but it's probably a matter of taste .. 

My issue was to be consistent with ListConnection and StreamConnection - they are directly handling the pagination parameters, assuming the DataFetcher does not (the arguments were anyway hidden to the DataFetcher, and the DataFetcher did not know what the cursors look like) - and here it will be the role of the DataFetcher to handle them. But I agree that it makes more sense to let the DataFetcher handle the pagination and return a simple interface as you suggest. And if we get rid of ListConnection and StreamConnection, we don't have to worry about being consistent anymore. Or if we want to keep them, we can simplify them to not handle the pagination at all, and let the DataFetcher do it.

By the way, it will even be easier that PaginatedList inherits from Iterable (so only add `hasNextPage hasPreviousPage getCursor`, `data` would actually be `iterator()` from `Iterable`). That would avoid creating a new type function (IterableFunction would handle it), and it will allow the same type to work with or without the @GraphQLConnection annotation.

**Comment by guy120494** [comment]
> Or if we want to keep them, we can simplify them to not handle the pagination at all, and let the DataFetcher do it

I dont see how we can do that, because SimpleListConnection assumes that all the data is returned, and builds the sub list from that (doesn't matter whether or not the dataFetcher handled the pagination itself).

I think I'm just going to get rid of SimpleLIstConnection

**Comment by tdraier** [comment]
Yes, i was thinking of getting rid of SimpleListConnection and have a simpler implementation that would handle list.

**Comment by guy120494** [comment]
Actually it was not so easy to get it done, because I figured out that now we can't allow connection on fields (because the PropertyDataFetcher doesn't know how to handle connection), but I got it done

File: src/main/java/graphql/annotations/GraphQLAnnotations.java
Comment (yarinvak) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java]: its perferct for a seperate method

File: src/main/java/graphql/annotations/GraphQLAnnotations.java
Comment (yarinvak) [inline src/main/java/graphql/annotations/GraphQLAnnotations.java]: it throws an exception? why not to handle it here?

**Review by yarinvak — APPROVED** [review]
Awesome

File: src/main/java/graphql/annotations/GraphQLAnnotations.java
Comment (guy120494) [reply src/main/java/graphql/annotations/GraphQLAnnotations.java]: I want the exception to be thrown, because it suggest that the `@GraphQLConnection` is not used properly, so it must be fixed in order to build the schema correclty

File: src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java]: Typo in class name ( Abstarct )

File: src/main/java/graphql/annotations/connection/ConnectionTypeValidator.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/ConnectionTypeValidator.java]: From my understanding, PaginatedData is only required for "EnhancedConnectionFetcher", but we could perfectly imagine that somebody implement another ConnectionFetcher not based on PaginatedData. Actually if someone really wants to have the same kind of stuff as the previous SimpleListConnection based on a List, the validator should not block it. I think the validator class should actually be provided by the EnhancedConnectionFetcher itself (either through an annotation on EnhancedConnectionFetcher, or through a static method ?)

File: src/main/java/graphql/annotations/connection/ConnectionDataFetcher.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/ConnectionDataFetcher.java]: PaginatedData is tied to EnhancedConnectionFetcher, but here we should not assume we have DataFetcher<PaginatedData<T>> , only a DataFetcher<?> that would be understood by the ConnectionFetcher implementation. (also see comment below)

File: src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java]: Could rename it "PaginatedDataConnectionFetcher" ?

File: README.md
Comment (tdraier) [inline README.md]: I think this limitation is quite ok, but just for information, what is the issue with Property/FieldDataFetcher ?

File: src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java
Comment (tdraier) [inline src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java]: Maybe instead (or in addition to ?) of an AbstractPaginatedData, we could have a SimplePaginatedData which also implements a basic cursor implementation, like a base64 encoded offset ( as in the previous StreamConnectionFetcher or SimpleListConnection) ?

File: README.md
Comment (guy120494) [reply README.md]: Well, they don't know how to build connection (because it's not an Scalar type)

File: src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java]: I think that the abstract class helps building your own PaginatedData class.
I will add SimplePaginatedData

File: src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/AbstarctPaginatedData.java]: done

File: src/main/java/graphql/annotations/connection/ConnectionTypeValidator.java
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/ConnectionTypeValidator.java]: now this class implements an interface called ConnectionValidator, and it can be overridden
in EnhancedConnectionFetcher (now called PaginatedDataConnectionFetcher)

File: src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java
Comment (guy120494) [reply src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java]: done

**Comment by tdraier** [comment]
Thanks, everything looks good to me, I will try to use it as soon as possible ! For me it's ok to merge.

**Review by tdraier — APPROVED** [review]

**Comment by guy120494** [comment]
I want to wait until we merge #122, and then we will merge this PR.
If you can also code review #122 it would be great!

**Pull request merged**
